### DISCUSSION TITLE
Why not consider implementing a stackless coroutine?

### DISCUSSION THREAD

**Discussion by OuYangPaste** [edubart/nelua-lang#172] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/38561133/149789110-4e925ead-2e72-4f36-99c9-14bae70dfd90.png)
uclibc does not support ucontext,  i am sad

**Comment by edubart** [comment, marked as answer]
Coroutines in Nelua are designed like in Lua, so they are stackful as in Lua, and there are no plans to support stackless coroutines in the compiler. This is a controversial topic but I believe stackful coroutines are more powerful, useful and easier to use, although it comes with some disadvantages that we are aware of, like it may use more memory footprint and it requires proper support for the target machine, usually via some assembly coding, thus it is not cross platform as stackless.

Seems like you are compiling for MIPS architecture, the Nelua coroutine implementation is backed by the [minicoro](https://github.com/edubart/minicoro) project, it provides stackful coroutines for C, it was also created by myself and currently it does not support MIPS yet, but it could. To do the MIPS implementation it requires some MIPS assembly coding, PRs are welcome there. The `ucontext` implementation is used only as fallback implementation in Unix systems, however `ucontext` is deprecated and heavier, thus it is not recommended to rely on it, what is really lacking here is the proper assembly implementation for MIPS architecture in the minicoro project.

**Reply by lePereT** [reply]
Hey edubart, do you have any plans to introduce growable coroutine stacks in minicoro/nelua? This would allow for the use of very many lightweight coroutines so that we can implement delightful concurrency frameworks like fibers (https://wingolog.org/archives/2018/05/16/lightweight-concurrency-in-lua for an example)

**Reply by edubart** [reply]
> Hey edubart, do you have any plans to introduce growable coroutine stacks in minicoro/nelua?

No plans, coroutines with growable stacks would need to map new stack pages using the OS virtual memory APIs and deal with page faults, just to keep references of anything in the stack valid when growing. This would kill portability, introduce a lot of unwanted complexity, and make the implementation dependent in the OS. Right now the implementation is more dependent on the architecture, not much in the OS, and that makes minicoro attractive, more portable and easier to maintain.

Nevertheless, you could use thousands of small coroutines with fixed stack size, you just need to know in advance the maximum stack usage for each coroutine and use that value when creating a new coroutine, just take care to not use a value so small that would lead to stack overflows. If the default coroutine stack is too big for your use case, just adjust it.

**Comment by OuYangPaste** [comment]
Is it possible that it looks like stackful in nelua but is stackless when it is translated into C, such as dividing it into multiple C functions with **yield** ?

**Reply by OuYangPaste** [reply]
I thought again. Before calling a function in a stackless coroutine, it must be marked, like async, await, which is not like lua.
